### PR TITLE
Add change affected input to fingerprint

### PR DESCRIPTION
**Pull request by norahuang**
[AB#1613487](https://dev.azure.com/mseng/708e929f-6bd5-415a-8daf-25b1dac08dd8/_workitems/edit/1613487)

### CODE REVIEW

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs [hunk 1]
```diff
@@ -50,9 +50,15 @@ public abstract class PipFingerprinter
...
 
+        /// <summary>
+        /// Refers to a function which maps a process to its source change affected dependencies. 
+        /// </summary>
+        public delegate IReadOnlyList<AbsolutePath> SourceChangeAffectedContentsLookup(Process process);
+
         private readonly PathTable m_pathTable;
         private readonly PipFragmentRenderer.ContentHashLookup m_contentHashLookup;
         private readonly PipDataLookup m_pipDataLookup;
+        private readonly SourceChangeAffectedContentsLookup m_sourceChangeAffectedContentsLookup;
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: It's not just sources. It's basically for all inputs to process.

**Comment by narasamdya** [comment]
>     public abstract class PipFingerprinter

Can you add a test for `PipFingerprinterTest.cs` ?

---
Refers to: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs:40 in 8b4702c. [](commit_id = 8b4702cb5d5a693a076771d287d4b172f4a938a2, deletion_comment = False)

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs [hunk 2]
```diff
@@ -50,9 +50,15 @@ public abstract class PipFingerprinter
         /// </summary>
         public delegate PipData PipDataLookup(FileArtifact artifact);
 
+        /// <summary>
+        /// Refers to a function which maps a process to its source change affected dependencies.
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: >source [](start = 61, length = 6)

its inputs -- sources are source files, but inputs can be produced by another process.

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs [hunk 3]
```diff
@@ -336,6 +344,11 @@ protected virtual void AddWeakFingerprint(IFingerprinter fingerprinter, Process
             }
 
             fingerprinter.AddCollection<int, ReadOnlyArray<int>>("SuccessExitCodes", process.SuccessExitCodes, (h, i) => h.Add(i));
+
+            if (process.ChangeAffectedInputListWrittenFilePath.IsValid)
```
Comment (narasamdya) [inline Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: Why is the path `ChangeAffectedInputListWrittenFilePath` is not part of the fingerprint? What if I change that path?

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 4]
```diff
@@ -74,6 +74,93 @@ public void DirectAffectedFileInputTest()
...
+
+            // Build1 with change in aInput
+            File.WriteAllText(changeList.Path.ToString(Context.PathTable), aInput.Path.ToString(Context.PathTable));
+            var result = RunScheduler();
+            result.AssertCacheMiss(pipA.Process.PipId, pipB.Process.PipId, pipC.Process.PipId);
+            var actualAffectedInput = File.ReadAllText(changeAffectedWrittenFile.ToString(Context.PathTable));
+            var expectedAffectedInput = aOutput.Path.GetName(Context.PathTable).ToString(Context.PathTable.StringTable);
+            XAssert.AreEqual(expectedAffectedInput, actualAffectedInput);
+
+            ResetPipGraphBuilder();
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >ResetPipGraphBuilder [](start = 12, length = 20)

Why do you have to `ResetPipGraphBuilder`?

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 5]
```diff
@@ -74,6 +74,93 @@ public void DirectAffectedFileInputTest()
...
+                Operation.ReadFile(aOutput),
+                Operation.ReadFile(bOutput),
+                Operation.WriteFile(cOutput)
+            });
+            pipBuilderC.SetChangeAffectedInputListWrittenFile(changeAffectedWrittenFile);
+            var pipC = SchedulePipBuilder(pipBuilderC);
+            Configuration.Schedule.InputChanges = changeList.Path;
+
+            // Build1 with change in aInput
+            File.WriteAllText(changeList.Path.ToString(Context.PathTable), aInput.Path.ToString(Context.PathTable));
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >aInput.Path.ToStrin [](start = 75, length = 19)

I think we have a method called `ArtifactToString` in integration tests. Please review all the unit tests in this class.

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 6]
```diff
@@ -74,6 +74,93 @@ public void DirectAffectedFileInputTest()
...
+                Operation.ReadFile(FileArtifact.CreateSourceFile(changeAffectedWrittenFile), doNotInfer: true),
+                Operation.ReadFile(aOutput),
+                Operation.ReadFile(bOutput),
+                Operation.WriteFile(cOutput)
+            });
+            pipBuilderC.SetChangeAffectedInputListWrittenFile(changeAffectedWrittenFile);
+            var pipC = SchedulePipBuilder(pipBuilderC);
+            Configuration.Schedule.InputChanges = changeList.Path;
+
+            // Build1 with change in aInput
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >// Build1 with change in aInput [](start = 12, length = 31)

Run Build0 where you don't have any change.

**Comment by narasamdya** [comment]
>     public class ChangeAffectedInputTests : SchedulerIntegrationTestBase

This class should also be tested in incremental scheduling mode. See the incrementalschedulingtests folder in integration test project.

---
Refers to: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs:19 in 8b4702c. [](commit_id = 8b4702cb5d5a693a076771d287d4b172f4a938a2, deletion_comment = False)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 7]
```diff
@@ -74,6 +74,93 @@ public void DirectAffectedFileInputTest()
...
+            // Build3 revert change in bInput
+            File.WriteAllText(bInput.Path.ToString(Context.PathTable), "pipBOrigin");
+            File.WriteAllText(changeList.Path.ToString(Context.PathTable), bInput.Path.ToString(Context.PathTable));
+            pipA = SchedulePipBuilder(pipBuilderA);
+            pipB = SchedulePipBuilder(pipBuilderB);
+            pipC = SchedulePipBuilder(pipBuilderC);
+
+            result = RunScheduler();
+            result.AssertCacheHit(pipA.Process.PipId, pipB.Process.PipId);
+            result.AssertCacheMiss(pipC.Process.PipId);
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >result.AssertCacheMiss(pipC.Process.PipId); [](start = 12, length = 43)

Give comment why you expect to have cache miss here.

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs (hunk 1 above)
Comment (norahuang) [reply Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: Source change affected is a attribute defining contents.

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs (hunk 2 above)
Comment (norahuang) [reply Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: Let me add - in between the words. source-change-affected

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs (hunk 3 above)
Comment (norahuang) [reply Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: This is an internal used file. We never defined it as a input or output file. Thurs, its location shouldn't matters and shouldn't be in the fingerprint.

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs (hunk 5 above)
Comment (norahuang) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: I uses ArtifactToString for all other cases, except this one. I will change it.

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs (hunk 4 above)
Comment (norahuang) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: It make sense to me to reset pip graph builder for a new build.

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs (hunk 6 above)
Comment (norahuang) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: Added

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs (hunk 7 above)
Comment (norahuang) [reply Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: will do

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/IncrementalSchedulingTests/ChangeAffectedInputTests_IncrementalScheduling.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+{
+    [Feature(Features.IncrementalScheduling)]
+    [TestClassIfSupported(requiresJournalScan: true)]
+    public class ChangeAffectedInputTests_IncrementalScheduling : ChangeAffectedInputTests
+    {
+        public ChangeAffectedInputTests_IncrementalScheduling(ITestOutputHelper output) : base(output)
+        {
+            Configuration.Schedule.IncrementalScheduling = true;
+            Configuration.Schedule.SkipHashSourceFile = false;
+        }
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/IncrementalSchedulingTests/ChangeAffectedInputTests_IncrementalScheduling.cs]: >} [](start = 8, length = 1)

nit: Remove new line.

File: Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs (hunk 3 above)
Comment (narasamdya) [reply Public/Src/Engine/Scheduler/Fingerprints/PipFingerprinter.cs]: However, you allow users to set its location, don't you? If that's the case, I think the path should go to the fingerprint as well.

---
In reply to: [336684057](https://github.com/microsoft/BuildXL/pull/1064#discussion_r336684057) [](ancestors = 336684057)

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 9]
```diff
@@ -43,45 +43,129 @@ public void DirectAffectedFileInputTest()
...
-            SchedulePipBuilder(pipBuilderB);
+            var pipB = SchedulePipBuilder(pipBuilderB);
 
             var inputChangesFile = CreateOutputFileArtifact();
             File.WriteAllText(ArtifactToString(inputChangesFile), ArtifactToString(aInput));
             Configuration.Schedule.InputChanges = inputChangesFile.Path;
 
-            RunScheduler().AssertCacheMiss();
+            // changeAffectedWrittenFile of pipB contains aOutput, expecting cache miss
+            RunScheduler().AssertCacheMiss(pipB.Process.PipId);
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >RunScheduler().AssertCacheMiss(pipB.Process.PipId); [](start = 12, length = 51)

Can you assert that pipA will be a cache hit?

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 10]
```diff
@@ -43,45 +43,129 @@ public void DirectAffectedFileInputTest()
...
+
+            result = RunScheduler();
+            result.AssertCacheHit(pipA.Process.PipId);
+            result.AssertCacheMiss(pipB.Process.PipId, pipC.Process.PipId);
+
+            actualAffectedInput = File.ReadAllText(changeAffectedWrittenFile.ToString(Context.PathTable));
+            expectedAffectedInput = bOutput.Path.GetName(Context.PathTable).ToString(Context.PathTable.StringTable);
+            XAssert.AreEqual(expectedAffectedInput, actualAffectedInput);
+
+            // Build3 revert change in bInput. pipB get cache hit of build0.
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: typo: reverts

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 11]
```diff
@@ -161,31 +245,35 @@ public void TransitiveAffectedDirectoryInputTest(InputAccessType pipBInputAccess
...
 
+            // Creating an input dir for pipA
             var aInputDir = Path.Combine(ObjectRoot, "input");
             var aInputDirPath = AbsolutePath.Create(Context.PathTable, aInputDir);
             var aInputDirArtifact = DirectoryArtifact.CreateWithZeroPartialSealId(aInputDirPath);
             var aInputFile = CreateSourceFile(root: aInputDirPath, prefix: "pip-a-input-file");
             File.WriteAllText(ArtifactToString(aInputFile), "pipABuild1");
             var sealInputDir = SealDirectory(aInputDirPath, SealDirectoryKind.SourceAllDirectories);
 
+            // Creating an output dir Artifact(containing a outfile and a sub output dir) for pipB
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >( [](start = 46, length = 1)

nit: space

File: Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs [hunk 12]
```diff
@@ -337,7 +435,7 @@ public void AffectedSourceInsideSourceSealDirectory()
             var changeAffectedWrittenFile = CreateUniqueObjPath("change");
             var pipBuilder = CreatePipBuilder(new Operation[]
             {
-                // Ensure that pip reads the changeAffectedWrittenFile.
+                // Ensure that pip reads the changeAffectedWrittenFile if it exist.
```
Comment (narasamdya) [inline Public/Src/Engine/UnitTests/Scheduler.IntegrationTest/ChangeAffectedInputTests.cs]: >exist [](start = 77, length = 5)

nit: exists

**Pull request merged**